base/compiler/inferencestate.jl

```diff
@@ -488,7 +506,68 @@ function compute_trycatch(code::Vector{Any}, ip::BitSet, bbs::Union{Vector{Basic
     end
 
     @assert first(ip) == n + 1
-    return handler_at, handlers
+    return handler_info
+end
+
+function is_throw_call(e::Expr, code::Vector{Any})
+    if e.head === :call
+        f = e.args[1]
+        if isa(f, SSAValue)
+            f = code[f.id]
+        end
+        if isa(f, GlobalRef)
+            ff = abstract_eval_globalref_type(f)
+            if isa(ff, Const) && ff.val === Core.throw
+                return true
+            end
+        end
+    end
+    return false
+end
+
+function mark_throw_blocks!(src::CodeInfo, handler_info::Union{Nothing,HandlerInfo})
+    for stmt in find_throw_blocks(src.code, handler_info)
+        src.ssaflags[stmt] |= IR_FLAG_THROW_BLOCK
+    end
+    return nothing
+end
+
+function find_throw_blocks(code::Vector{Any}, handler_info::Union{Nothing,HandlerInfo})
```

Comment:
Orthogonal to this PR since this code isn't new, but it might be worth adding a note that this under-estimates the must-throw paths (both because of how it traverses the stmts and also because it only notices `throw()` calls), so it's only useful as a heuristic.